### PR TITLE
MINOR: improve Streams checkstyle and code cleanup

### PR DESCRIPTION
**Pull request by mjsax**
While fixing the code to meet the new checkstyle rule, I updated all touched classed for Java8 and removed other warnings.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java [hunk 1]
```diff
@@ -76,22 +76,25 @@
     /** Maximum time difference for tuples that are after the join tuple. */
     public final long afterMs;
 
-    private final Duration grace;
+    private final long graceMs;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:80]: This might be controversial. I change it, to reuse the return type of `ApiUtils.validateMillisecondDuration`. (Similar in other XxxWindows classes).

Thoughts?

**Comment by mjsax** [comment]
Call for review @guozhangwang @bbejeck @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java:80]: It is fine to me.

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java [hunk 2]
```diff
@@ -72,7 +72,7 @@
  * }</pre>
  */
 @InterfaceStability.Evolving
-public class Stores {
+public final class Stores {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: Is this intentional? I'm concerned someone out there in the wild may have extended this class for their customized needs.

File: streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java [hunk 3]
```diff
@@ -68,23 +62,24 @@
...
+        String adID;
 
         /* other attributes */
-        public long eventTime;
-        public String userID = UUID.randomUUID().toString(); // not used
-        public String pageID = UUID.randomUUID().toString(); // not used
-        public String addType = "banner78";  // not used
-        public String ipAddress = "1.2.3.4"; // not used
+        long eventTime;
+        //public String userID = UUID.randomUUID().toString(); // not used
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java]: Is this intentional? If they are retained as comments just for context information I'd suggest use `/**/` as a nit comment.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStoreTest.java [hunk 4]
```diff
@@ -127,7 +127,7 @@ public void shouldThrowInvalidStateStoreExceptionOnRebalance() {
 
     @Test
     public void shouldThrowInvalidStateStoreExceptionIfFetchThrows() {
-        underlyingWindowStore.setOpen(false);
+        underlyingWindowStore.setOpenToFalse();
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyWindowStoreTest.java]: Ditto here.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 5]
```diff
@@ -213,11 +211,11 @@ public boolean isValidTransition(final State newState) {
     private final Object stateLock = new Object();
     protected volatile State state = State.CREATED;
 
-    private boolean waitOnState(final State targetState, final long waitMs) {
+    private boolean waitOnNoRunningState(final long waitMs) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I'd prefer keep it as general so that the same function can be reused in the future potentially rather than explicitly restrict is usage scope.

**Review by guozhangwang — APPROVED** [review]
Feel free to merge after addressed / replied the comments.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 6]
```diff
@@ -919,9 +911,7 @@ public void run() {
      * @throws IllegalArgumentException if {@code timeout} can't be represented as {@code long milliseconds}
      */
     public synchronized boolean close(final Duration timeout) throws IllegalArgumentException {
-        ApiUtils.validateMillisecondDuration(timeout, "timeout");
-
-        final long timeoutMs = timeout.toMillis();
+        final long timeoutMs = ApiUtils.validateMillisecondDuration(timeout, "timeout");
```
Comment (nizhikov) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I think `validateXXX` method shouldn't return value.
That's why I made it void in the first place.

Maybe we need a more convenient name for it?

File: streams/src/main/java/org/apache/kafka/streams/state/Stores.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/Stores.java]: There was a warning. Note that `Stores` has only static methods, thus it does not make sense to inherit from it. I can still revert it. Don't feel strong about it.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 6 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: > That's why I made it void in the first place.

It's not `void` -- and this PR does not change it. I think it makes sense to let it return for this case though. To validate the given `Duration`, we need to call `#toMillis()` within `validateMillisecondDuration()` and thus, there is no reason to call it twice (even if this is not on the critical code path).

We can still rename it. Maybe something like, `getMillisecondDurationOrFailIfInvalid` ?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 7]
```diff
@@ -810,6 +804,7 @@ public void run() {
      * Shutdown this {@code KafkaStreams} instance by signaling all the threads to stop, and then wait for them to join.
      * This will block until all threads have stopped.
      */
+    @SuppressWarnings("deprecation")
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Why is this needed?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java [hunk 8]
```diff
@@ -160,12 +160,12 @@ public Cancellable schedule(final long interval, final PunctuationType type, fin
         return task.schedule(interval, type, callback);
     }
 
+    @SuppressWarnings("deprecation")
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java:181]: Again, why are we adding calls to deprecated code?

File: checkstyle/checkstyle.xml [hunk 9]
```diff
@@ -77,6 +77,12 @@
       <property name="validateEnhancedForLoopVariable" value="true"/>
     </module>
 
+    <!-- if/else/for/while/do blocks should use braces (suppressed except for Streams) -->
+    <module name="NeedBraces">
+      <property name="tokens" value="LITERAL_IF, LITERAL_ELSE, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO"/>
```
Comment (ijuma) [inline checkstyle/checkstyle.xml]: It's a real shame that we have this inconsistency. And it only makes the code worse, this PR being a good example.

File: streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java [hunk 10]
```diff
@@ -68,23 +62,26 @@
...
+        String adID;
 
         /* other attributes */
-        public long eventTime;
-        public String userID = UUID.randomUUID().toString(); // not used
-        public String pageID = UUID.randomUUID().toString(); // not used
-        public String addType = "banner78";  // not used
-        public String ipAddress = "1.2.3.4"; // not used
+        long eventTime;
+        /* not used
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java:70]: Should we just remove these?

**Review by bbejeck — APPROVED** [review]
Just one minor question, otherwise LGTM modulo other comments

File: checkstyle/checkstyle.xml (hunk 9 above)
Comment (mjsax) [reply checkstyle/checkstyle.xml]: I am happy to enable this check style for the whole code base (similar to `final` requirement). But not sure if we get some push back -- thus, it seems to be better to introduce those things step by step?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 7 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Because we call deprecated method `close(DEFAULT_CLOSE_TIMEOUT, TimeUnit.SECONDS);` -- this method will stay though, ie, we will make it private when we "remove" it from public API and than we can remove this annotation, too.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java (hunk 8 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java:181]: We don't add calls -- this call was always there and we need to keep it for backward compatibility until we remove those methods (by making them private).

File: streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java (hunk 10 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java:70]: I don't have strong opinion -- might be nice to keep for documentation (ie, the benchmark describes those fields, but we just don't use them?) Thoughts?

File: checkstyle/checkstyle.xml (hunk 9 above)
Comment (ijuma) [reply checkstyle/checkstyle.xml]: I was suggesting the opposite. What's the justification for this rule? At least for `final`, there is the mutability argument. The truth is that Java is becoming a more modern language and redundant verbosity is being eliminated (an extreme example that has been proposed http://openjdk.java.net/jeps/8209434). Requiring braces in every situation perhaps made sense for the old Java. :)

Comment (mjsax) [reply checkstyle/checkstyle.xml]: For me personally, it's readability question -- I think, it improves readability.

Also, this rule only required braces for if/else and loop-blocks. Not for _every_ situation.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 11]
```diff
@@ -811,7 +804,7 @@ public void run() {
      * This will block until all threads have stopped.
      */
     public void close() {
-        close(DEFAULT_CLOSE_TIMEOUT, TimeUnit.SECONDS);
+        close(Long.MAX_VALUE);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:810]: @ijuma Removed the `SuppressWarning` annotation and rewrote the code. Also have a PR for 2.1 branch: https://github.com/apache/kafka/pull/5963 for this fix.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:810]: We translate old default of zero to Long.MAX_VALUE within deprecated `close(final long timeout, final TimeUnit timeUnit)` -- we can call `private close()` directly instead.

File: streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java (hunk 10 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/perf/YahooBenchmark.java:70]: That makes sense.  Maybe we could add a couple of words in the comment? 
But it's not a big deal, so it could just stay as is.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 11 above)
Comment (ijuma) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:810]: Thanks, this looks better.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/KafkaEmbedded.java [hunk 12]
```diff
@@ -186,6 +187,7 @@ public void createTopic(final String topic,
         }
     }
 
+    @SuppressWarnings("WeakerAccess")
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/KafkaEmbedded.java:190]: Are these IntelliJ warnings?

File: checkstyle/checkstyle.xml (hunk 9 above)
Comment (ijuma) [reply checkstyle/checkstyle.xml]: Hmm, what other situation is there where braces are optional?

Comment (mjsax) [reply checkstyle/checkstyle.xml]: switch-case statement

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/KafkaEmbedded.java (hunk 12 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/KafkaEmbedded.java:190]: Yes.

File: checkstyle/checkstyle.xml (hunk 9 above)
Comment (mjsax) [reply checkstyle/checkstyle.xml]: http://checkstyle.sourceforge.net/config_blocks.html#NeedBraces

**Comment by mjsax** [comment]
Rebased this. Also added some JavaDoc fixes.

@ijuma Do you still have objections about the enforced use of braces in streams module?

**Comment by ijuma** [comment]
@mjsax I'm not going to block it, but this doesn't make sense to me. The Apache Kafka project is a single project and it should use the same code style across the board for a given programming language, in my opinion. 

Imagine if we had a different code style per module: clients, connect, streams, tools. Do you think that would work OK? Why is streams special?

**Comment by guozhangwang** [comment]
I think I'd agree with @ijuma 's concern that, it is not for arguing if which coding style should be preferred, but about maintaining the coding style consistency of modules under AK repo.

Personally I also prefer unambiguous syntax and I see the merits of, for example, posing strict reviews on proper markups / javadocs on public APIs etc in practice, but enforcing checkstyle rules only for Streams may have potential side-effect to the integral community in long run (I've seen other projects leading to this direction that resulted in separated sub-projects). So I'd suggest we remove the checkstyle rule while merging this PR, and moving forward we would still encourage reviewers to use their own best judgement to determine if some places would be vulnerable to future bugs or really bad for readability that worth a nitpick comment.

**Comment by mjsax** [comment]
Removed checkstyle rule. Merged all other cleanups. Thanks for your comments and the discussion about it!

Merged to `trunk`
